libpcsxcore/lightrec/plugin.c

```diff
@@ -171,6 +171,30 @@ static void lightrec_tansition_from_pcsx(struct lightrec_state *state)
 	}
 }
 
+static void reset_timer_data(struct lightrec_state *state, unsigned int timer)
+{
+	lightrec_set_timer_data(state, timer,
+				cycles_pcsx_to_lightrec(rcnts[timer].cycleStart),
+				rcnts[timer].rate);
+}
+
+static void maybe_reset_timer_data(struct lightrec_state *state, u32 addr)
+{
+	switch (addr & 0x1fffffff) {
+	case 0x1f801100:
+	case 0x1f801104:
+	case 0x1f801108:
+	case 0x1f801110:
+	case 0x1f801114:
+	case 0x1f801118:
+	case 0x1f801120:
+	case 0x1f801124:
+	case 0x1f801128:
+		reset_timer_data(state, (addr & 0x30) >> 4);
+		break;
+	}
+}
```

Comment:
I think this can only work for timer 2 really. 0 and 1 have modes synced to the GPU where the timer is is reset on each scanline. Those are pretty much unemulated right now and will be needed for lightgun game support.

And does this work in general? Lightrec still counts 2 cycles/instruction, so the timer will be based on that? Does Parasite Eve II work with this?

Also what about about wrapping? Since psxRegs.cycle runs slower that lightrec's cycle counter, as soon as soon as it wraps won't there be a large timer jump observable by the game? For example with default settings the ratio is something like 1.14, so psxRegs.cycle counter value of ~0 ~= 613M lightrec's counter, then it wraps and cycles_pcsx_to_lightrec() does as well causing sudden jump to close to 0.

Reply:
Yes, wrapping is probably an issue. The reason I can't reset the counter every time is that the timers' `cycleStart` value is below psxRegs.cycle. But I can for instance reset Lightrec's cycle counter to e.g. 0x8000'0000 every time, and compute the "cycleStart" value that's passed to Lightrec accordingly. What matters are the offsets and not really the real values.

I did not know that timers 0 / 1 should reset on each scanline, but as you said, currently in PCSX it doesn't. If timers 0 / 1 need to reset on each scanline, I think it's still possible to handle it without jumping to C, using a modulo.

Right now the behaviour (except wrapping etc.) of Lightrec should be the exact same as calling the `psxRcntRcount` C function... in theory. When I was working on this optimization I still had a version of pcsx_rearmed without the CPU ratio implemented for Lightrec, and it matched 100%. Now with the cycles ratio implemented I'm struggling a bit (hence the RFC). I just tested Parasite Eve 2, and it does not work with this change. I guess because Lightrec still counts 2 cycles / instruction, yes.

One solution would probably be to use a 22.10 cycle counter in Lightrec, with a configurable cycles/op ratio in 22.10 format as well. Then the CPU ratio wouldn't have to be handled in the glue in the first place.

Btw, one problem about the CPU ratio approach is the loss of accuracy. Converting one value from PCSX to Lightrec scale then back to PCSX won't give you the exact same number.